File: Py/PDA Py/BalancedParantheses.py
import time

stack = []
timeInitial = time.time()
f = open("C:\\Users\\Kavita manoj\\OneDrive\\Desktop\\PDA Py\\tests\\51")
expression = f.read()
for x in expression:
    stack.append(x)
top = -1

for x in expression:
    if x == '[' or x == '(' or x == '{':
        stack.append(x)
        top+=1
    elif stack[top] == '[' and x == ']':
        stack.pop()
        top-=1
    elif stack [top] == '(' and x == ')':
        stack.pop()
        top-=1
    elif stack[top] == '{' and x == '}':
        stack.pop()
        top-=1

if top == -1:
    print ("Balanced")
else:
    print ("Unbalanced")

timeFinal = time.time()
print ("Time taken = ", timeFinal-timeInitial)
print ("for total string length", len(expression))

File: Sem 3-4-5-6 rep/C++ VSCode/DSA Codes Rev/Stack/BalancedParantheses.cpp
#include <iostream>
#include <string>
using namespace std;
char stack[100];
int top = -1;

void push(char element){
    if (top == 100){
        cout << "Overflow";
    }
    else{
    top++;
    stack[top] = element;
    }
}

void pop (){
    if (top == -1){
        cout << "Underflow" << endl;
    }
    else {
        top--;
    }
}

int main(){
    string expression;
    char c;
    int len = expression.length();
    cout << "Enter an expression: ";
    getline(cin, expression);
    for (int i = 0; i < len; i++){
        c = expression.at(i);
        if (c == '[' || c == '{' || c == '('){
            push(c);
        }
        else if (c == ']' && stack[top] == '['){
            pop();
        }
        else if (c == '}' && stack[top] == '{'){
            pop();
        }
        else if (c == ')' && stack[top] == '('){
            pop();
        }
    }
    if (top==-1){
        
        cout << "Balanced." << endl;
    }
    else {
        cout << "Unbalanced." << endl;
    }

}